gap/SnAnUnknownDegree.gi

```diff
@@ -0,0 +1,345 @@
+        if nrTriedConjugates >= C or nrThreeCycleCandidates >= T then
+            r := RandomElm(ri,"simplesocle",true)!.el;
+            a := 0;
+            tPower := r ^ M;
+            # Invariant: tPower = (r ^ M) ^ (2 ^ a)
+            repeat
+                a := a + 1;
+                tPowerOld := tPower;
+                tPower := tPower ^ 2;
+            until a = logInt2N or isone(ri)(tPower);
+            if a = logInt2N then
+                return NeverApplicable;
+            fi;
+            t := tPowerOld;
+            nrInvolutions := nrInvolutions + 1;
+            nrTriedConjugates := 0;
+            nrThreeCycleCandidates := 0;
+        fi;
+        # Steps 4 & 5: new three cycle candidate
+        # Try to construct a three cycle candidate via a conjugate of t. See
+        # the comment above this function.
+        nrTriedConjugates := nrTriedConjugates + 1;
+        c := t ^ RandomElm(ri,"simplesocle",true)!.el;
+        if not isequal(ri)(t * c, c * t) then
+            nrThreeCycleCandidates := nrThreeCycleCandidates + 1;
+            return (t * c) ^ 2;
+        else
+            return fail;
+        fi;
+    end;
+    # construct the iterator
+    oneThreeCycleCandidate := function()
+        local candidate;
+        repeat
+            if nrInvolutions >= B
+                and (nrTriedConjugates >= C or nrThreeCycleCandidates >= T)
+            then
+                # We are done and were not able to recognize Sn or An.
+                return TemporaryFailure;
+            fi;
+            candidate := tryThreeCycleCandidate();
+            if candidate = NeverApplicable then
+                return NeverApplicable;
+            fi;
+        until candidate <> fail;
+        return candidate;
+    end;
+    return oneThreeCycleCandidate;
+end);
+
+# G: the group to recognize
+# c: possibly a 3-cycle
+# returns a list of group elements. If G is isomorphic to an alternating or
+# symmetric group and c is a 3-cycle, then this function returns a list of
+# bolstering elements with respect to c.
+BindGlobal("BolsteringElements",
+function(ri, c, eps, N)
+    local result, R, S, prebolsteringElms, i, r, cr, cr2;
+    result := [];
+    R := Int(Ceil(7 / 4 * Log2(Float(eps ^ -1))));
+    S := 7 * N * R;
+    prebolsteringElms := [];
+    i := 0;
+    # find pre-bolstering elements
+    while i <= S and Length(prebolsteringElms) <= R do
+        r := RandomElm(ri,"simplesocle",true)!.el;
+        # test whether r is pre-bolstering
+        cr := c ^ r;
+        cr2 := c ^ (r ^ 2);
+        if not isone(ri)(Comm(cr, c))
+                and not isequal(ri)(cr2, c)
+                and not isequal(ri)(cr2, c ^ 2)
+                and isone(ri)(Comm(cr2, c))
+        then
+            Add(prebolsteringElms, r);
+        fi;
+        i := i + 1;
+    od;
+    # construct bolstering elements
+    for r in prebolsteringElms do
+        if isone(ri)((c ^ (r * c * r)
+                      * c ^ (r * c ^ (r ^ 2) * c)) ^ 3)
```

Comment:
I think it would more sense to merge this into the preceeding loop: instead of sticking `r` into `prebolsteringElms` and then getting it out again, computing some more, and sticking the results into `result`; just move the processing code below up and replace the `Add(prebolsteringElms, r);` code line by it (and then adjust the condition of the surrounding loop, and remove `prebolsteringElms`.

What you gained by this is that you have `cr` and `cr2` available
```suggestion
        if isone(ri)((cr ^ (c * r)
                      * cr ^ (cr2 * c)) ^ 3)
```

Of course in the end, we need to produce SLPs anyway, and perhaps they are clever enough to figure out some of this, but I am not sure I'd want to rely on that?

Reply:
Done